### ISSUE TITLE
Checkpointing with matching time windows

### ISSUE THREAD

**Issue by DavidSCN** [precice/openfoam-adapter#210]
We have oftentimes the case that we reload the checkpoint after a single time step was computed (we have matching time windows in an implicit coupling). In the current implementation of the adapter and the way we handle the checkpointing, we copy all fields and reload them later

https://github.com/precice/openfoam-adapter/blob/17d1776a31f96958c5629d0af948675028667f9b/Adapter.C#L858-L882

This might be somewhat wasteful in case OpenFOAM stores fields of the previous time step automatically. If it does, we could access these 'old' fields (of t_n-1) in order to reload the checkpoint (t_n -> t_n-1) instead of storing both snapshots explicitly (t_n and t_n-1) and reloading them later on (would need to reload the old fields, though). This could save us a considerable amount of memory (namely all global fields at least once).

However, I don't know, if the OpenFOAM fields at t_n-1 are stored and which we need to reload at all.

Extracted from https://github.com/precice/openfoam-adapter/pull/209#issuecomment-992652605.

**Comment by MakisH** [comment]
> We have oftentimes the case that we reload the checkpoint after a single time step was computed (we have matching time windows in an implicit coupling).

I assume you refer here to the case of no subcycling. We don't always need to have `solver_dt` = `precice_dt` in implicit coupling and the solver may do multiple time steps in one coupling time window.

How would you imagine this in terms of configuration? I guess we cannot really extract this information automatically, since OpenFOAM is allowed to arbitrarily update its time step size (and the user is allowed to update it at runtime).

**Comment by DavidSCN** [comment]
> How would you imagine this in terms of configuration? I guess we cannot really extract this information automatically, since OpenFOAM is allowed to arbitrarily update its time step size (and the user is allowed to update it at runtime).

Variable time step sizes as well as run time modifications can be disabled in the configuration file.

**Comment by MakisH** [comment]
> Variable time step sizes as well as run time modifications can be disabled in the configuration file.

Of course, but then do you mean that we should make it "if `solver_dt == precice_dt` and disabled runtime modifications and disabled dynamic timestep size, then use the alternative mechanism"? We would then probably need to ensure that the `precice_dt` we are comparing with is the maximum one (e.g., we can compare during initialization).

**Comment by DavidSCN** [comment]
Actually, I did thought about configuring this yet. Let's first see if there is anything we can improve on before bothering with the details.
